libdrgn/stack_trace.c

```diff
@@ -1130,6 +1130,37 @@ drgn_program_stack_trace(struct drgn_program *prog, uint32_t tid,
 	return drgn_get_stack_trace(prog, tid, NULL, NULL, ret);
 }
 
+LIBDRGN_PUBLIC struct drgn_error *
+drgn_program_stack_trace_from_pcs(struct drgn_program *prog, uint64_t *pcs,
+				  size_t pcs_size, struct drgn_stack_trace **ret)
+{
+	struct drgn_stack_trace *trace =
+		malloc(offsetof(struct drgn_stack_trace, frames[pcs_size]));
+	struct drgn_error *err;
+	size_t trace_capacity = pcs_size;
+
+	if (!trace)
+		return &drgn_enomem;
+
+	trace->prog = prog;
+	trace->num_frames = 0;
+	for (size_t i = 0; i != pcs_size; ++i) {
+		struct drgn_register_state *regs =
+			drgn_register_state_create_impl(0, 0, false);
```

Comment:
Setting `interrupted` to false means that this PC is the return address of some call, not the instruction itself that trapped or whatever. Often, the top frame is interrupted (e.g., in a core dump), but not always (e.g., for a scheduled out task). We probably want this interface to allow controlling the `interrupted` flag for the top frame.

Do you know whether stack frames in the stack depot store a return address or an interrupted address in the top frame?

Reply:
> Do you know whether stack frames in the stack depot store a return address or an interrupted address in the top frame?

At least for the current users of stack depot I think they're all return addresses. E.g. almost every caller of `stack_depot_save` or `__stack_depot_save` calls `stack_trace_save` to collect the stack trace (the exception is KMSAN which uses the stack depot to store partial stack traces in a non-standard format), which calls `arch_stack_walk` with a null `pt_regs` argument, which only ever passes return addresses to the callback (at least the arm64 and x86 implementations do).

> We probably want this interface to allow controlling the interrupted flag for the top frame.

Is it worth doing that now? Personally I'd wait until there is a user.

Reply:
Fair enough, it'll be easy enough to add later.